File: mycollab-ui/src/main/java/com/mycollab/vaadin/mvp/AbstractSingleContainerPageView.java
/**
 * This file is part of mycollab-ui.
 *
 * mycollab-ui is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-ui is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-ui.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.vaadin.mvp;

import com.vaadin.ui.Component;
import com.vaadin.ui.CssLayout;
import com.vaadin.ui.CustomComponent;
import com.vaadin.ui.SingleComponentContainer;

/**
 * @author MyCollab Ltd
 * @since 5.4.5
 */
public class AbstractSingleContainerPageView extends CustomComponent implements PageView, SingleComponentContainer {

    private CssLayout contentLayout;

    public AbstractSingleContainerPageView() {
        contentLayout = new CssLayout();
        contentLayout.setSizeFull();
        setCompositionRoot(contentLayout);
        setSizeFull();
    }

    @Override
    public Component getContent() {
        return contentLayout.getComponent(0);
    }

    @Override
    public void setContent(Component component) {
        contentLayout.removeAllComponents();
        contentLayout.addComponent(component);
    }

    @Override
    public void addComponentAttachListener(ComponentAttachListener componentAttachListener) {
        contentLayout.addComponentAttachListener(componentAttachListener);
    }

    @Override
    public void removeComponentAttachListener(ComponentAttachListener componentAttachListener) {

    }

    @Override
    public void addComponentDetachListener(ComponentDetachListener componentDetachListener) {

    }

    @Override
    public void removeComponentDetachListener(ComponentDetachListener componentDetachListener) {

    }

    @Override
    public void addViewListener(ViewListener listener) {
        addListener(ViewEvent.VIEW_IDENTIFIER(), ViewEvent.class, listener, ViewListener.viewInitMethod);
    }
}

File: mycollab-ui/src/main/java/com/mycollab/vaadin/mvp/AbstractSingleContainerPageView.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.vaadin.mvp

import com.mycollab.vaadin.event.ViewEvent
import com.vaadin.ui.*
import com.vaadin.ui.HasComponents.ComponentAttachListener
import com.vaadin.ui.HasComponents.ComponentDetachListener

/**
 * @author MyCollab Ltd
 * @since 5.4.5
 */
open class AbstractSingleContainerPageView : CustomComponent(), PageView, SingleComponentContainer {

    private val contentLayout: CssLayout = CssLayout()

    init {
        contentLayout.setSizeFull()
        compositionRoot = contentLayout
        setSizeFull()
    }

    override fun getContent(): Component {
        return contentLayout.getComponent(0)
    }

    override fun setContent(component: Component) {
        contentLayout.removeAllComponents()
        contentLayout.addComponent(component)
    }

    override fun addComponentAttachListener(componentAttachListener: ComponentAttachListener) {
        contentLayout.addComponentAttachListener(componentAttachListener)
    }

    override fun removeComponentAttachListener(componentAttachListener: ComponentAttachListener) {

    }

    override fun addComponentDetachListener(componentDetachListener: ComponentDetachListener) {

    }

    override fun removeComponentDetachListener(componentDetachListener: ComponentDetachListener) {

    }

    override fun <E> addViewListener(listener: PageView.ViewListener<E>) {
        addListener(ViewEvent.VIEW_IDENTIFIER, ViewEvent::class.java, listener, PageView.ViewListener.viewInitMethod)
    }
}
